docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,331 @@
+        - However, since the mechanism for retrieving endorsement keys varies across platforms, the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key.
+
+#### Provider Specification
+
+##### All providers
+- All providers MUST insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names MUST be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner
+- All providers SHOULD implement trust-on-first-use type mechanisms to prevent replay attacks
+
+##### Insecure providers
+- An insecure provider CANNOT implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer
+
+##### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+##### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm`
+    - `cluster-x-k8s-io/kubelet-serving-tpm`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation.
+- Kubeadm COULD allow opt-out of kubeadm setting up `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+This is deferred to a future Kubeadm design and release, and for this proposal, we will add fields to KubeadmControlPlane to remove these node groups and bindings post control plane initialisation.
+
+#### Changes to KubeadmControlPlane
+
+```
+spec:
+  disableKubeadmAutoApproveClusterRoleBinding: true
```

Comment:
We should probably find a better name for this field, were any alternatives considered?

@fabriziopandini Thoughts here?

Reply:
I personally prefer to not add fields to the KubeadmConfig root, because it is already a mix of kubeadm config and cloud-init/machine-bootstrap fields and I don't think we should add further concerns on the same bucket.

What about adding a `KubeadmConfig.Security` field, pointing to a struct holding all the security related settings? 

Also, I'm wondering if we should have something like `security.nodeAttestation` instead of `security. disableKubeadmAutoApproveClusterRoleBinding`, so the user will be concerned only to express the final intent, leaving to the system the job to figure out what is required to make it happen.

Reply:
> were any alternatives considered?

Added it mostly as a placeholder that there _should_ be a field of some sort.

Reply:
Changed it to the following with justification.

```
spec:
  security:
    disableKubeadmAutoApproveClusterRoleBinding: true
```

This field on KCP will remove the `kubeadm:node-autoapprove-bootstrap` cluster role binding.

We do not name the field as related to the Cluster API specific mechanism here since this
setting would also support the use of other mechanisms such as the aws-iam-authenticator or the
GCP cloud provider.